### ISSUE TITLE
Start Publishing Staging Images to Artifact Registry

### ISSUE THREAD

**Issue by upodroid** [kubernetes/k8s.io#3961]
Part of: #1343

Notes: https://gist.github.com/upodroid/a33723a7e1abc5e9c6fabc6b07e7aac0

When images are built after a PR is merged, they need to be pushed to Artifact Registry(AR).

~Unlike Google Container Registry(GCR), you don't need a separate GCP project per staging project. The permission boundary is at the repository level instead of the project. This allows multiple registries to be created per project. Therefore, we will create a new project, start publishing images to it and delete the other projects after the transition.~

~We will need to do the following:
    - create a new GCP project (k8s-artifacts-staging)
    - create a docker image repository per project
    - modify the jobs to push images to prow.
    - Provision the infra with terraform. There is a separate issue open to deploy changes to terraform via Prow on merges.~

Prod changes need to be done via the shell scripts which makes deploying staging via terraform kind of pointless.

Open questions:
- Do we need to backfill the staging repository via gcrane? 

@puerco 

/area artifacts
/priority important-soon
/area release-eng

**Comment by upodroid** [comment]
After #3968 is merged, we need to do a few things:
- repos that use ko to build images(k-sigs/bom,etc), modify KO_DOCKER_REPO variable to `us-docker.pkg.dev/k8s-staging-bom/images/bom` as an example
- Modify image-builder to tweak the cloudbuild.yaml file

Bit unsure about other things.

**Comment by upodroid** [comment]
Spoke about this at the sig-testing meeting. Will push it out to the future when GCR deprecation is actually announced by Google.

_May want to standardize on Aritfact Registry eventually, but would require tracking down a number of staging registries and changing team push processes (with unclear payoff right now)_

/priority backlog
/sig testing

**Comment by BenTheElder** [comment]
Strawman:

1) Create AR versions of all staging registries. These can be empty, but should have the same IAM access / names as the GCRs
2) Ask subprojects to switch to the AR registry. They do so by:
   1. Switching new builds to push to AR (this will vary by how they're doing image push)
   2. Running a backfill task we provide (`gcrane cp -R`)
   3. Switching from GCR to AR in image promoter config

We should probably check in with release engineering about that last part.

We could alternately:
1) Create AR versions of all staging registries.
2) SIG K8s Infra spins up a job or jobs to continuously backfill from GCR to AR (`grane cp -R` ... potentially excessive API usage ...)
3) SIG K8s Infra swaps over all the promoter manifests to promote from AR
4) We ask subprojects to stop writing to GCR so we can wind down the grane sync for their repo.

The first plan is less resource intensive but more people-coordination heavy

**Comment by ameukam** [comment]
Another alternative could be to transition existing GCR repos to AR repos: https://cloud.google.com/artifact-registry/docs/transition/setup-gcr-repo

The infrastructure changes would be minimal and there is little to be done by the community. The downsize of this approach is mostly cost. we will use multi-regional AR repositories.

**Comment by BenTheElder** [comment]
I forgot about this! Nice! We should definitely estimate the cost delta of that approach ...

One other downside: skew versus newly created registries going forward.

**Comment by ameukam** [comment]
> One other downside: skew versus newly created registries going forward.

AR allow us to create AR repos using `gcr.io` (traffic is redirected from `gcr.io` to the AR repo) so I think we are fine until GCR is gone.

**Comment by BenTheElder** [comment]
Maybe instead tracking in https://github.com/kubernetes/k8s.io/issues/1343

**Comment by upodroid** [comment]
Not rotten

**Comment by dims** [comment]
Need to go back and create `etcd-manager` staging bucket in https://github.com/kubernetes/k8s.io/pull/6897

slack conversation : https://kubernetes.slack.com/archives/CCK68P2Q2/p1718654806739639

**Comment by BenTheElder** [comment]
(Some projects are on AR, but not all, thanks to @upodroid)

**Comment by ameukam** [comment]
We have exceptions for migration:
 - https://github.com/kubernetes/k8s.io/blob/main/infra/gcp/terraform/k8s-staging-images/registries.tf#L35-L41

**Comment by upodroid** [comment]
Quick update, I'm on track to finish the ACTIVE projects by the deadline by Google(18th of March)

Outstanding projects:

```
# some of these have been moved to k8s-staging-images AR instances such as #7824
 mahamed  Mahameds-MacBook-Pro  ~  1  USAGE  $  gcloud container images list-gcr-usage     --folder=713040427754 --filter="usage=ACTIVE"
---
repository: gcr.io/k8s-staging-lws
usage: ACTIVE
---
repository: gcr.io/k8s-staging-jobset
usage: ACTIVE
---
repository: gcr.io/k8s-staging-provider-os
usage: ACTIVE
---
repository: gcr.io/k8s-staging-tejolote
usage: ACTIVE
---
repository: gcr.io/k8s-staging-zeitgeist
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kmm
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kwok
usage: ACTIVE
---
repository: gcr.io/k8s-staging-win-op-rdnss
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kueue
usage: ACTIVE
---
repository: gcr.io/k8s-staging-gmsa-webhook
usage: ACTIVE
---
repository: gcr.io/k8s-staging-sched-simulator
usage: ACTIVE
---
repository: gcr.io/k8s-staging-prometheus-adapter
usage: ACTIVE
---
repository: gcr.io/k8s-staging-gateway-api
usage: ACTIVE
---
repository: gcr.io/k8s-staging-releng-test
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kustomize
usage: ACTIVE
---
repository: gcr.io/k8s-staging-networking
usage: ACTIVE
---
repository: gcr.io/k8s-staging-sp-operator
usage: ACTIVE
---
repository: gcr.io/k8s-staging-sig-docs
usage: ACTIVE
---
repository: gcr.io/k8s-staging-scheduler-plugins
usage: ACTIVE
---
repository: gcr.io/k8s-staging-git-sync
usage: ACTIVE
---
repository: gcr.io/k8s-staging-sig-storage
usage: ACTIVE
---
repository: gcr.io/k8s-staging-ingressconformance
usage: ACTIVE
---
repository: gcr.io/k8s-staging-ingress-nginx
usage: ACTIVE
---
repository: gcr.io/k8s-staging-provider-aws
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kube-state-metrics
usage: ACTIVE
---
repository: gcr.io/k8s-staging-slack-infra
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kubernetes
usage: ACTIVE
---
repository: gcr.io/k8s-staging-scl-image-builder
usage: ACTIVE
---
repository: gcr.io/k8s-staging-nfd
usage: ACTIVE
---
repository: gcr.io/k8s-staging-npd
usage: ACTIVE
---
repository: gcr.io/k8s-staging-external-dns
usage: ACTIVE
---
repository: gcr.io/k8s-staging-provider-azure
usage: ACTIVE
---
repository: gcr.io/k8s-staging-kas-network-proxy
usage: ACTIVE
---
repository: gcr.io/k8s-staging-multitenancy
usage: ACTIVE
---
repository: gcr.io/k8s-staging-publishing-bot
usage: ACTIVE
```

Successfully Migrated Projects Projects:

```
 mahamed  Mahameds-MacBook-Pro  ~  1  $  gcloud container images list-gcr-usage     --folder=713040427754 --filter="usage=REDIRECTED"
---
repository: gcr.io/k8s-staging-cloud-pv-vsphere
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-ipam-ic
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cloud-pv-labeler
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api-helm
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-operator
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-cloudstack
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-bom
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cloud-provider-ibm
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-ibmcloud
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api-nested
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-test-infra
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cri-tools
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-ci-images
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-addon-manager
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-addons
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cloud-provider-gcp
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-boskos
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-infra-tools
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cpa
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-etcd
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-csi-secrets-store
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-dns
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api-do
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-autoscaling
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-vsphere
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-apisnoop
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-e2e-test-images
usage: REDIRECTED
---
repository: asia.gcr.io/k8s-staging-cip-test
usage: REDIRECTED
---
repository: eu.gcr.io/k8s-staging-cip-test
usage: REDIRECTED
---
repository: us.gcr.io/k8s-staging-cip-test
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api-azure
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api-gcp
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-docker
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-openstack
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-artifact-promoter
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-capi-kubeadm
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-descheduler
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-build-image
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api-aws
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cip-test
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-coredns
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-csi
usage: REDIRECTED
---
repository: gcr.io/k8s-staging-cluster-api
usage: REDIRECTED
```

Projects with inactive GCR usage. They will be migrated in late March after the ACTIVE projects are done.

```
 mahamed  Mahameds-MacBook-Pro  ~  1  $  gcloud container images list-gcr-usage     --folder=713040427754 --filter="usage=INACTIVE"
---
repository: gcr.io/k8s-staging-wasm-scheduler
usage: INACTIVE
---
repository: gcr.io/k8s-staging-kubebuilder
usage: INACTIVE
---
repository: gcr.io/k8s-staging-win-svc-proxy
usage: INACTIVE
---
repository: gcr.io/k8s-staging-tg-exporter
usage: INACTIVE
---
repository: gcr.io/k8s-staging-krm-functions
usage: INACTIVE
---
repository: gcr.io/k8s-staging-sig-auth
usage: INACTIVE
---
repository: gcr.io/k8s-staging-kubetest2
usage: INACTIVE
---
repository: gcr.io/k8s-staging-provider-openstack
usage: INACTIVE
---
repository: gcr.io/k8s-staging-experimental
usage: INACTIVE
---
repository: gcr.io/k8s-staging-examples
usage: INACTIVE
---
repository: gcr.io/k8s-staging-storage-migrator
usage: INACTIVE
---
repository: gcr.io/k8s-staging-etcdadm
usage: INACTIVE
---
repository: gcr.io/k8s-staging-k8s-gsm-tools
usage: INACTIVE
---
repository: gcr.io/k8s-staging-kubeadm
usage: INACTIVE
```

**Comment by ameukam** [comment]
@upodroid Are we not done with this ?

**Comment by BenTheElder** [comment]
@thockin TLDR: yes git-sync was moved and you don't need to do anything.

----

x-ref https://github.com/etcd-io/etcd/issues/15199#issuecomment-2637803054
```shell
# check for legacy GCR, which used storage.googleapis.com
crane pull --verbose gcr.io/k8s-staging-git-sync/git-sync:v3.6.9 /dev/null 2>&1 | grep storage
# nothing

# check for artifact registry
$ crane pull --verbose gcr.io/k8s-staging-git-sync/git-sync:v3.6.9 /dev/null 2>&1 | grep /artifacts-downloads | head -n1
Location: /artifacts-downloads/namespaces/k8s-staging-git-sync/repositories/gcr.io/downloads/[snipped super long hash url]
```

You, as an owner of it, can also browse to gcr.io/k8s-staging-git-sync and see that it redirects you to `console.cloud.google.com/artifacts/docker/k8s-staging-git-sync/us/gcr.io`

You don't need to do anything, there's some loose plans to migrate to AR domains, but it's not essential, for now it is migrated via https://cloud.google.com/artifact-registry/docs/transition/auto-migrate-gcr-ar and you don't need to do anything.

**Comment by BenTheElder** [comment]
I think we can close this.
We can separately track attempts to shift to AR domains directly.
